# Upload Data from Dave's Redistricting

Due to discrepancies in 538's data and untraceable sourcing, I have pulled together public data from Dave's Redistricting App on 3.20.2022.

At the point of initial pull, Missouri, Louisiana, New Hampshire, and Florida still had incomplete maps. 

Also, Kentucky and West Virginia, for some reason, did not yet have data for the 2020 Presidential election yet. Ohio's new map was also not included. For these states, we'll still have to use 538's less traceable data.

In [1]:
import requests
import pandas as pd
import os
import glob
import numpy as np
import math
pd.set_option("display.max_rows", None, "display.max_columns", None)

In [2]:
#load the files into a single dataframe
path = "/Users/xavier/Desktop/DSPP/solo_projects/redistricting_project/raw_data/daves_pvis"
all_files = sorted(glob.glob(path + "/*.csv"))

li = []

for filename in all_files:
    df = pd.read_csv(filename, index_col=None, header=0).reset_index()
    df["ST"] = filename[-6:-4]
    li.append(df)

frame = pd.concat(li, axis=0, ignore_index=False)

In [3]:
#cut the frame and fix the columns
df = frame[["ST","index","Devation","Dem"]]
df.columns = ["ST","district","dem_pct","gop_pct"]
#remove nonumeric rows
possible_numbers = list([str(i) for i in range(53)])
df = df[df["district"].isin(possible_numbers)]
df.shape

(362, 4)

Without Florida, Ohio, Missouri, Kentucky, Louisiana, New Hampshire having valid data, and the 6 single district states not being in the set, 362 is the exact number we should have

# Clean the Dave's data into our standardized format

In [4]:
#Input standard text values
df["year"] = "2022"
df["congress"] = "118"
df["ST#"] = df["ST"] + df["district"]
#compute metric
df["raw_metric"] = (.50 + df["gop_pct"] - .4831) * 100
df["round_metric"] = df["raw_metric"].apply(np.around)
df["metric"] = round(df["raw_metric"]/100,2)
#compute PVI
df["lean"] = np.where(df["metric"] > .5, "R", "D")
df['PVI'] = df['lean'] + "+" + ((((df['round_metric']-50)).abs())).fillna(1000).astype(int).astype(str)
df['PVI'] = df['PVI'].replace("D+0","EVEN")
#df['PVI'] = np.where(df['PVI'].str[-1]=="+","EVEN",df['PVI'])
#df["metric"] = df['PVI'].str.split("+").str[1]
#df['metric'] = np.where(df['PVI']=="EVEN","50",df['metric'])
#df["metric"] = df["metric"].astype(int)
#df["metric"] = np.where(df["lean"]=="D", (df["metric"]/100))
#checker = df[["raw_metric","round_metric","metric","PVI"]]
#checker

In [5]:
#df[abs(df["round_metric"] - df["raw_metric"]) > .4]
df

ST district  dem_pct  gop_pct  year congress   ST#  raw_metric  \
1   AL        1   0.3535   0.6465  2022      118   AL1       66.34   
2   AL        2   0.3423   0.6577  2022      118   AL2       67.46   
3   AL        3   0.3248   0.6752  2022      118   AL3       69.21   
4   AL        4   0.1870   0.8130  2022      118   AL4       82.99   
5   AL        5   0.3457   0.6543  2022      118   AL5       67.12   
6   AL        6   0.3253   0.6747  2022      118   AL6       69.16   
7   AL        7   0.6596   0.3404  2022      118   AL7       35.73   
1   AR        1   0.3025   0.6975  2022      118   AR1       71.44   
2   AR        2   0.4267   0.5733  2022      118   AR2       59.02   
3   AR        3   0.3638   0.6362  2022      118   AR3       65.31   
4   AR        4   0.3295   0.6705  2022      118   AR4       68.74   
1   AZ        1   0.4943   0.5057  2022      118   AZ1       52.26   
2   AZ        2   0.4537   0.5463  2022      118   AZ2       56.32   
3   AZ        3   0.7619   0.2381  2022      118   AZ3       25.50   
4   AZ        4   0.5359   0.4641  2022      118   AZ4       48.10   
5   AZ        5   0.4039   0.5961  2022      118   AZ5       61.30   
6   AZ        6   0.4879   0.5121  2022      118   AZ6       52.90   
7   AZ        7   0.6721   0.3279  2022      118   AZ7       34.48   
8   AZ        8   0.4173   0.5827  2022      118   AZ8       59.96   
9   AZ        9   0.3597   0.6403  2022      118   AZ9       65.72   
1   CA        1   0.3937   0.6063  2022      118   CA1       62.32   
2   CA        2   0.7502   0.2498  2022      118   CA2       26.67   
3   CA        3   0.4767   0.5233  2022      118   CA3       54.02   
4   CA        4   0.6863   0.3137  2022      118   CA4       33.06   
5   CA        5   0.4292   0.5708  2022      118   CA5       58.77   
6   CA        6   0.5930   0.4070  2022      118   CA6       42.39   
7   CA        7   0.6965   0.3035  2022      118   CA7       32.04   
8   CA        8   0.7843   0.2157  2022      118   CA8       23.26   
9   CA        9   0.5664   0.4336  2022      118   CA9       45.05   
10  CA       10   0.6974   0.3026  2022      118  CA10       31.95   
11  CA       11   0.8926   0.1074  2022      118  CA11       12.43   
12  CA       12   0.9208   0.0792  2022      118  CA12        9.61   
13  CA       13   0.5601   0.4399  2022      118  CA13       45.68   
14  CA       14   0.7403   0.2597  2022      118  CA14       27.66   
15  CA       15   0.8011   0.1989  2022      118  CA15       21.58   
16  CA       16   0.7803   0.2197  2022      118  CA16       23.66   
17  CA       17   0.7610   0.2390  2022      118  CA17       25.59   
18  CA       18   0.7398   0.2602  2022      118  CA18       27.71   
19  CA       19   0.7060   0.2940  2022      118  CA19       31.09   
20  CA       20   0.3575   0.6425  2022      118  CA20       65.94   
21  CA       21   0.6087   0.3913  2022      118  CA21       40.82   
22  CA       22   0.5738   0.4262  2022      118  CA22       44.31   
23  CA       23   0.4376   0.5624  2022      118  CA23       57.93   
24  CA       24   0.6427   0.3573  2022      118  CA24       37.42   
25  CA       25   0.5839   0.4161  2022      118  CA25       43.30   
26  CA       26   0.5975   0.4025  2022      118  CA26       41.94   
27  CA       27   0.5583   0.4417  2022      118  CA27       45.86   
28  CA       28   0.6771   0.3229  2022      118  CA28       33.98   
29  CA       29   0.7889   0.2111  2022      118  CA29       22.80   
30  CA       30   0.7550   0.2450  2022      118  CA30       26.19   
31  CA       31   0.6750   0.3250  2022      118  CA31       34.19   
32  CA       32   0.7208   0.2792  2022      118  CA32       29.61   
33  CA       33   0.6418   0.3582  2022      118  CA33       37.51   
34  CA       34   0.8556   0.1444  2022      118  CA34       16.13   
35  CA       35   0.6596   0.3404  2022      118  CA35       35.73   
36  CA       36   0.7301   0.2699  2022      118  CA36       28.68   
37  CA       3

In [6]:
#create a clean version of the data to match other outputs
pvi_118 = df[["year","congress","ST","ST#","PVI","metric"]]

# Load in 538 Data to fill in gaps

In [7]:
#load in filepaths for 538's data
path_538 = "/Users/xavier/Desktop/DSPP/solo_projects/redistricting_project/clean_data/supplimental/data_118_538.csv"
path_pre = "/Users/xavier/Desktop/DSPP/solo_projects/redistricting_project/clean_data/supplimental/previews_118.csv"

In [8]:
#grab states with maps not available in Dave's
no_data_states = ["KY","WV","OH","VT", "DE", "WY", "ND", "SD", "AK"]
load_538 = pd.read_csv(path_538)
no_data_rows = load_538[load_538["ST"].isin(no_data_states)]
#pull in maps not yet passed
pre_rows = pd.read_csv(path_pre)

In [9]:
pvi_118 = pd.concat([pvi_118,no_data_rows,pre_rows]).sort_values("ST")
pvi_118.shape

(435, 6)

Now we have something to represent all 435 Districts, we can export a dataset with the best possible estimates of what the 2022 map will look like

In [10]:
pvi_118.to_csv("/Users/xavier/Desktop/DSPP/solo_projects/redistricting_project/clean_data/full_districts/data_118.csv",index=False)

In [11]:
pvi_118[pvi_118["metric"].isna()]

Empty DataFrame
Columns: [year, congress, ST, ST#, PVI, metric]
Index: []